// @serverless/cloud/src/models/Data.js
'use strict'

const request = require('../utils/request')

class Data {
  constructor(data = {}) {
    this.fs = data.fs
    this.api = data.api
    this.org = data.org
    this.service = data.service
    this.instance = data.instance
  }

  async export() {
    const { dataFileName } = this.fs

    const options = {
      accessKey: this.org.accessKey,
      method: 'GET',
      endpoint: `/orgs/${this.org.orgName}/services/${this.service.serviceName}/instances/${this.instance.instanceName}/data`,
      params: {
        dataFileName
      }
    }

    const data = await request(options)

    await this.fs.writeDataFile(data)
  }

  async import(overwrite) {
    const { dataFileName } = this.fs

    const items = await this.fs.readDataFile()

    const options = {
      accessKey: this.org.accessKey,
      method: 'PUT',
      endpoint: `/orgs/${this.org.orgName}/services/${this.service.serviceName}/instances/${this.instance.instanceName}/data`,
      body: {
        items,
        overwrite
      },
      params: {
        dataFileName
      }
    }

    await request(options)
  }
}

module.exports = Data
